### PR TITLE
Fixing mypy issues with strict_optional enabled.

### LINKED ISSUE #25: Add static type checker to tests
**Is your feature request related to a problem? Please describe.**
We would like to have a high consistency in our code.

**Describe the solution you'd like**
Add static type checker mypy: https://mypy.readthedocs.io/en/latest/

**Describe alternatives you've considered**
None.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR addresses the issues raised when mypy is enabled with `strict_optional`.

## Fixes

If fixes #25 

## Types of changes

What types of changes does your code introduce to agents-tac?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

This is WIP.

### CODE REVIEW

File: aea/aea.py [hunk 1]
```diff
@@ -55,9 +55,7 @@ def __init__(self, name: str,
         super().__init__(name=name, private_key_pem_path=private_key_pem_path, timeout=timeout, debug=debug)
 
         self.max_reactions = max_reactions
-        self._directory = directory
-        if self._directory is None:
-            self._directory = str(Path(".").absolute())
+        self._directory = directory if directory is not None else str(Path(".").absolute())
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply aea/aea.py]: OK, so let's put `directory: str = ""` as an argument, and then 

```python
         self._directory = directory if directory else str(Path(".").absolute())
```

File: aea/aea.py [hunk 2]
```diff
@@ -76,6 +74,7 @@ def act(self) -> None:
 
         :return: None
         """
+        assert self.resources is not None, "Call setup before calling act."
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply aea/aea.py]: we could solve this problem by requiring that `self.resource` must be initialized in the constructor. That implies an `AEA` must always have a resource directory. If not specified, initialize the resource directory with defaults.

File: aea/agent.py [hunk 3]
```diff
@@ -80,14 +80,16 @@ def __init__(self, name: str,
         self.mailbox = None  # type: Optional[MailBox]
 
     @property
-    def inbox(self) -> Optional[InBox]:
+    def inbox(self) -> InBox:
         """Get the inbox."""
-        return self.mailbox.inbox if self.mailbox else None
+        assert self.mailbox is not None, "Cannot retrieve inbox. No mailbox specified."
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply aea/agent.py]: I think you're doing well by adding these assert statements. What we want is to be MyPy compliant, then we can clean the code step by step.

File: aea/channel/oef.py [hunk 4]
```diff
@@ -477,7 +474,7 @@ def connect(self) -> None:
         if self._stopped and not self._connected:
             self._stopped = False
             self._core.run_threaded()
-            assert self.channel.connect()
+            assert self.channel.connect(), "Cannot connect to OEFChannel."
```
Comment (marcofavorito) [inline aea/channel/oef.py]: I think we should remove this `assert` statement. eventually. And we should override the OEF SDK agent methods for connection succeeded/failed/terminated.

File: aea/mail/base.py [hunk 5]
```diff
@@ -36,10 +36,10 @@
 class Envelope:
     """The top level message class."""
 
-    def __init__(self, to: Optional[Address] = None,
-                 sender: Optional[Address] = None,
-                 protocol_id: Optional[ProtocolId] = None,
-                 message: Optional[bytes] = b""):
+    def __init__(self, to: Address,
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply aea/mail/base.py]: Agreed.

File: aea/agent.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply aea/agent.py]: Agreed!

File: aea/channel/oef.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/channel/oef.py]: Agreed, will leave for now though.

File: aea/aea.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply aea/aea.py]: I solved this differently now via the assert statement in a getter. We can revert to yours but one argument against it is that it renders the `setup` method pointless which I think conceptually should be there.

File: aea/channel/oef.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/channel/oef.py]: Yes agreed.

**Pull request merged**